### PR TITLE
bash scripts are now migrated to yq v4

### LINKED ISSUE #202: [5] Migrate to yq v4
yq v4 is released which is quite different to v3 (it's more like jq now, see https://mikefarah.gitbook.io/yq/v/v4.x/upgrading-from-v3)

We either be very explicit in which version we support or we want to upgrade to match the latest and greatest.

### PR DESCRIPTION
**Pull request by Pavan-Gunda**
**What this PR does / why we need it**:
bash scripts are now migrated to yq v4
**Which issue this PR fixes** *(use the format `fixes #<issue number>(, fixes #<issue_number>, ...)` to automatically close the issue when PR gets merged)*: fixes  #202

**Public facing documentation PR** *(if applicable)*

**Special notes for reviewer**:

**Add a screenshot or an example to illustrate the proposed solution:**

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [ ] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [x] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [x] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [ ] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

**Comment by crssnd** [comment]
can we use `yq4` with the news scripts? and install yq4 along side yq3?

we should also be able to use the previous apps releases

**Comment by crssnd** [comment]
also add the new yq in https://github.com/elastisys/compliantkubernetes-apps/blob/main/get-requirements.yaml

**Comment by Pavan-Gunda** [comment]
> can we use `yq4` with the news scripts? and install yq4 along side yq3?
> 
> we should also be able to use the previous apps releases

I can't see an immediate answer to having two yq versions alongside.

**Comment by Pavan-Gunda** [comment]
> can we use `yq4` with the news scripts? and install yq4 along side yq3?
> 
> we should also be able to use the previous apps releases

I can't see an immediate answer to having two yq versions alongside.

**Comment by robinAwallace** [comment]
@aarnq you can now take a look at this PR :slightly_smiling_face:

File: bin/common.bash [hunk 1]
```diff
@@ -146,11 +152,12 @@ yq_copy_changes() {
     source2=$2
     target=$3
 
-    keys=$(yq read "${source2}" --tojson --printMode pv '**' | \
-           sed -rn 's/\{"(.+)":.*\}/\1/p' | sed -r 's/\.\[.+\].*//' | sed -r 's/\\//g' | uniq)
+    keys=$(yq4 '.. | select(tag != "!!seq" and tag != "!!map") | path |
+           with(.[] | select(contains(".")); . = "\"" + . + "\"") |
+           join(".")' "${source2}" | \
+           sed -r 's/\.[0-9].*//' | sed -r 's/\\//g' | uniq)
```
Comment (aarnq) [inline bin/common.bash:158]:
```suggestion
    keys=$(yq4 '.. | select(tag != "!!map") | path |
           with(.[]; . = ("\"" + .) + "\"" ) |
           join "."' "$source2" | \
           sed -r 's/\."[0-9]".*//' | uniq)
```
I think this can be simplified a bit. Not selecting on `!!seq` won't allow us to set `[]` to null an array which is a regression.
The same style can be used in `yq_copy_commons` and `yq_read_block`, maybe this should just be it's own function.

File: bin/common.bash [hunk 2]
```diff
@@ -269,7 +276,7 @@ validate_config() {
         options=$(yq_read_block "${template_config}" "set-me")
         maybe_exit="false"
         for opt in ${options}; do
-            compare=$(yq compare "${template_config}" "${merged_config}" "${opt}" -j -ppv | sed -rn 's/\+(.*)/\1/p' || true)
+            compare=$(diff <(yq4 -P "${opt}" "${template_config}") <(yq4 -P "${opt}" "${merged_config}") || true)
```
Comment (aarnq) [inline bin/common.bash:279]:
```suggestion
            compare=$(diff <(yq4 -oj "${opt}" "${template_config}") <(yq4 -oj "${opt}" "${merged_config}") || true)
```
To remove comments before comparing.

File: bin/common.bash [hunk 3]
```diff
@@ -146,11 +152,12 @@ yq_copy_changes() {
...
-    keys=$(yq read "${source2}" --tojson --printMode pv '**' | \
-           sed -rn 's/\{"(.+)":.*\}/\1/p' | sed -r 's/\.\[.+\].*//' | sed -r 's/\\//g' | uniq)
+    keys=$(yq4 '.. | select(tag != "!!seq" and tag != "!!map") | path |
+           with(.[] | select(contains(".")); . = "\"" + . + "\"") |
+           join(".")' "${source2}" | \
+           sed -r 's/\.[0-9].*//' | sed -r 's/\\//g' | uniq)
     for key in ${keys}; do
-        compare=$(yq compare "${source1}" "${source2}" --tojson --printMode pv "${key}" | \
-                  sed -rn 's/^\+(.*)/\1/p' || true)
+        compare=$(diff <(yq4 -P ".${key}" "${source1}" ) <(yq4 -P ".${key}" "${source2}" ) || true)
```
Comment (aarnq) [inline bin/common.bash:160]:
```suggestion
        compare=$(diff <(yq4 -oj ".${key}" "${source1}" ) <(yq4 -oj ".${key}" "${source2}" ) || true)
```
Same thing here.

File: bin/common.bash [hunk 4]
```diff
@@ -121,16 +126,17 @@ yq_copy_commons() {
...
+    keys=$(yq_merge "${source1}" "${source2}" | \
+           yq4 '.. | select(tag != "!!seq" and tag != "!!map") | path |
+           with(.[] | select(contains(".")); . = "\"" + . + "\"") |
+           join(".")' | \
+           sed -r 's/\.[0-9].*//' | \
            sed -r 's/\\//g' | \
            uniq)
     for key in ${keys}; do
-        compare=$(yq compare "${source1}" "${source2}" --tojson --printMode pv "${key}" || true)
+        compare=$(diff <(yq4 -P ".${key}" "${source1}" ) <(yq4 -P ".${key}" "${source2}" ) || true)
```
Comment (aarnq) [inline bin/common.bash:137]:
```suggestion
        compare=$(diff <(yq4 -oj ".${key}" "${source1}" ) <(yq4 -oj ".${key}" "${source2}" ) || true)
```
Same here.

File: bin/common.bash [hunk 5]
```diff
@@ -94,25 +96,28 @@ log_error() {
...
+             sed -r 's/\.[0-9].*//' | sed -r 's/\\//g' | uniq
 }
 
 # Copies a block from one file to another
 # Usage: yq_copy_block <source> <target> <key>
 yq_copy_block() {
-    yq read "${1}" "${3}" --tojson | \
-    yq prefix - "${3}" --tojson | \
-    yq merge "${2}" - --inplace --overwrite --arrays overwrite --prettyPrint
+    prefix=$(python3 python/src/ck8s_py.py -k "${3}")
```
Comment (aarnq) [inline bin/common.bash:117]:
```suggestion
    prefix=$(yq4 -n ".$3 | path | reverse | .[] as \$i ireduce(\".\"; \"{\\\"\" + \$i + \"\\\":\" + . + \"}\")")
```
If you don't want the addition of python.

File: scripts/restore-sync/README.md [hunk 6]
```diff
@@ -14,8 +14,8 @@ This will guide you through restoring a bucket from an off-site backup.
 - Else have sync configured, template the `rclone-sync-config` secret, and apply:
   ```bash
   ./bin/ck8s ops helmfile sc -l app=rclone-sync template | \
-    yq r -c -d '*' -D '[]' - | yq r - 'kind==Secret' | \
-    yq w - 'metadata.name' 'rclone-sync-restore-config' | \
+    yq4 r -c -d '*' -D '[]' - | yq4 r - 'kind==Secret' | \
+    yq4 w - 'metadata.name' 'rclone-sync-restore-config' | \
```
Comment (aarnq) [inline scripts/restore-sync/README.md:18]: These needs some update.

File: scripts/restore/README.md [hunk 7]
```diff
@@ -10,8 +10,8 @@ It can be installed normally.
 Set these variables for the restore job:
 ```bash
 export CK8S_CONFIG_PATH=<your config path>
-export S3_BUCKET=$(yq read "${CK8S_CONFIG_PATH}/defaults/sc-config.yaml" objectStorage.buckets.harbor)
-export S3_REGION_ENDPOINT=$(yq read "${CK8S_CONFIG_PATH}/common-config.yaml" objectStorage.s3.regionEndpoint)
+export S3_BUCKET=$(yq4 read "${CK8S_CONFIG_PATH}/defaults/sc-config.yaml" objectStorage.buckets.harbor)
+export S3_REGION_ENDPOINT=$(yq4 read "${CK8S_CONFIG_PATH}/common-config.yaml" objectStorage.s3.regionEndpoint)
```
Comment (aarnq) [inline scripts/restore/README.md:14]: These too.

**Review by aarnq — COMMENTED** [review]
Overall it seems to work well.

File: bin/common.bash [hunk 8]
```diff
@@ -94,25 +96,29 @@ log_error() {
...
     source=$1
     value=$2
-
-    yq read "${source}" --tojson --printMode p "**(.==${value})" | \
-             sed -r 's/\.\[.*\].*//' | sed -r 's/\\//g' | uniq
+    # shellcheck disable=SC2140
+    yq4 ".. | select(tag != \"!!map\" and . == \"${value}\") | path |
+    with(.[]; . = (\"\\\"\" + .) + \"\\\"\" ) |
+    \".\" + join \".\"" "${source}" | \
+             sed -r 's/\."[0-9]".*//' | sed -r 's/\\//g' | uniq
```
Comment (aarnq) [inline bin/common.bash:112]:
```suggestion
             sed -r 's/\."[0-9]+".*//' | sed -r 's/\\//g' | uniq
```

File: bin/common.bash [hunk 9]
```diff
@@ -121,16 +127,17 @@ yq_copy_commons() {
...
 
-    keys=$(yq_merge "${source1}" "${source2}" --tojson | \
-           yq read - --tojson --printMode pv '**' | \
-           sed -rn 's/\{"(.+)":.*\}/\1/p' | \
-           sed -r 's/\.\[.+\].*//' | \
+    keys=$(yq_merge "${source1}" "${source2}" | \
+           yq4 '.. | select(tag != "!!map") | path |
+           with(.[]; . = ("\"" + .) + "\"" ) |
+           join "."' | \
+           sed -r 's/\."[0-9]".*//' | \
```
Comment (aarnq) [inline bin/common.bash:134]:
```suggestion
           sed -r 's/\."[0-9]+".*//' | \
```

File: bin/common.bash [hunk 10]
```diff
@@ -146,11 +153,12 @@ yq_copy_changes() {
     source2=$2
     target=$3
 
-    keys=$(yq read "${source2}" --tojson --printMode pv '**' | \
-           sed -rn 's/\{"(.+)":.*\}/\1/p' | sed -r 's/\.\[.+\].*//' | sed -r 's/\\//g' | uniq)
+    keys=$(yq4 '.. | select(tag != "!!map") | path |
+           with(.[]; . = ("\"" + .) + "\"" ) |
+           join "."' "$source2" | \
+           sed -r 's/\."[0-9]".*//' | uniq)
```
Comment (aarnq) [inline bin/common.bash:159]:
```suggestion
           sed -r 's/\."[0-9]+".*//' | uniq)
```

File: pipeline/init.bats [hunk 11]
```diff
@@ -92,23 +92,22 @@ teardown() {
             CK8S_CONFIG_PATH="${config_2}" run ck8s init
             assert_success
 
-            run diff "${config_1}/defaults/common-config.yaml" "${config_2}/defaults/common-config.yaml"
+            run diff <(sort "${config_1}/defaults/common-config.yaml") <(sort "${config_2}/defaults/common-config.yaml")
```
Comment (aarnq) [inline pipeline/init.bats:95]: Is the sort really needed, shouldn't we strive for it to not introduce unnecessary changes instead?

Comment (robinAwallace) [reply pipeline/init.bats:95]: Yes that would be preferred, but Im unsure how to make yq4 not change order on things.

Comment (robinAwallace) [reply pipeline/init.bats:95]: Okay, great. Thought too try it out now. And no have not tested it with the new changes. Before when i did a reinit it moved some stuff. Will update the PR once i have tested it.

Comment (robinAwallace) [reply pipeline/init.bats]: It worked. Wonder what fixed it :sweat_smile:

**Review by aarnq — APPROVED** [review]

File: WIP-CHANGELOG.md [hunk 12]
```diff
@@ -11,6 +11,7 @@
 
 ### Changed
 - The Kubernetes status Grafana dashboard (new node filter, new graphs for CPU/Memory requests and limits per node, updated graphs for CPU/Memory usage/requests)
+- bash scripts are now migrated to yq-v4.26.1
```
Comment (Xartos) [inline WIP-CHANGELOG.md:14]: I think you should add this as a release note. And also point out that you need to have yq version 4 as an alias `yq4` if you're not using the `get-requirements.yaml` playbook

File: bin/common.bash [hunk 13]
```diff
@@ -39,6 +39,8 @@ config_template_path="${root_path}/config"
 scripts_path="${root_path}/scripts"
 # shellcheck disable=SC2034
 pipeline_path="${root_path}/pipeline"
+# shellcheck disable=SC2034
+python_path="${root_path}/python/src"
```
Comment (Xartos) [inline bin/common.bash:43]: Can't find that this is used anywhere

**Review by Xartos — APPROVED** [review]

File: bin/common.bash [hunk 14]
```diff
@@ -318,13 +324,13 @@ validate_sops_config() {
         exit 1
     fi
 
-    rule_count=$(yq r - --length creation_rules < "${sops_config}")
+    rule_count=$(yq4 'length' < "${sops_config}")
```
Comment (Xartos) [inline bin/common.bash:327]: Should be 
```suggestion
    rule_count=$(yq4 '.creation_rules | length' < "${sops_config}")
```

File: WIP-CHANGELOG.md [hunk 15]
```diff
@@ -1,6 +1,8 @@
 ### Release notes
 - In 1.7 the cert-manager API versions v1alpha2, v1alpha3, and v1beta1, have been removed from the custom resource definitions (CRDs).
 - In 1.8 the cert-manager will validate the spec.privateKey.rotationPolicy on Certificate resources. Valid options are Never and Always.
+- bash scripts are now migrated to yq-v4.26.1
+  - Requires `yq4` as a alias to yq v4. Installed via `get-requirements.yaml`.
```
Comment (Xartos) [inline WIP-CHANGELOG.md:5]: Stealing @OlleLarsson feedback from kubespray
```suggestion
  - Requires `yq4` as an alias to yq v4. Installed via `get-requirements.yaml`.
```
